### PR TITLE
Fix bag.map bug when function raises StopIteration  

### PR DESCRIPTION
**Pull request by jrbourbeau**
This PR addresses the issue brought up by @luizberti in #3545. They found that when a function that raises a `StopIteration` is being applied to a dask bag via the `Bag.map` method, a different result is given than if using the built-in Python `map` function. The following code snippet can be used to illustrate the problem. 

<details>
<summary> Code example: </summary>

```python 
from dask import bag
from dask.array.utils import assert_eq

seq = list(range(20))
print('seq = {}'.format(seq))

def func(i):
    if i == 11:
        raise StopIteration
    else:
        return i

# Using dask bag map
db = bag.from_sequence(seq, npartitions=7)
dask_map_func = db.map(func).compute()
print('dask_map_func = {}'.format(dask_map_func))

# Using builtin Python map
map_func = list(map(func, seq))

assert_eq(dask_map_func, map_func)

```

</details>

<br>

After some digging it looks like the problem is originating from

https://github.com/dask/dask/blob/e1c48e0c970aeb81ffeb58791db2ca3ce76fa846/dask/bag/core.py#L1707-L1708

Here when `f(*a)` raises a `StopIteration`, the `for`-loop for a given chunk is truncated, but all other chunks that don't raise a `StopIteration` are present in the result. 

The idea I'm proposing here is to check if a `StopIteration` is raised, and if so then truncate the result accordingly using the Bag `__dask_postcompute__` method. 
 

- [ ] Tests added / passed
- [ ] Passes `flake8 dask`

### CODE REVIEW

**Comment by jrbourbeau** [comment]
Just wanted to get some feedback from others to see if this seems like a reasonable solution

File: dask/bag/core.py [hunk 1]
```diff
@@ -215,6 +215,17 @@ def finalize(results):
         results = toolz.concat(results)
     if isinstance(results, Iterator):
         results = list(results)
+
+    r = []
```
Comment (jrbourbeau) [inline dask/bag/core.py]: This section can probably be done in a cleaner way

**Comment by jrbourbeau** [comment]
I suspect the [broken travis test](https://travis-ci.org/dask/dask/jobs/388066814) might just need a restart (getting errors messages like `Worker 'gw0' crashed while running`)

**Comment by mrocklin** [comment]
It might be useful to write a couple tests for what you think correct behavior should be here.  I suspect that that will make it easier for reviewers to understand what is going on here than the code itself.  I'm not sure I'm smart enough to see all of the possible results from just code changes like these.

**Comment by jrbourbeau** [comment]
> It might be useful to write a couple tests for what you think correct behavior should be here

You bring up a good point :wink:

Due to the unordered nature of dask bags, I don't think the solution I proposed here is appropriate. I think a reasonable solution would be to just raise a `RuntimeWarning` letting users know that a `StopIteration` was raised and this can lead to unexpected behavior.

**Pull request closed without merging**